clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2222,17 +2224,41 @@ namespace {
       const FunctionType *FuncTy = PointeeType->getAs<FunctionType>();
       assert(FuncTy);
       const FunctionProtoType *FuncProtoTy = FuncTy->getAs<FunctionProtoType>();
-      if (!FuncProtoTy)
+
+      // If the callee and arguments will not be traversed as part of the
+      // checking below, traverse them here.  This prevents TraverseStmt
+      // from needing to traverse the children of call expressions.
+      if (!FuncProtoTy) {
+        TraverseChildren(E, CSS, Facts, SE);
         return ResultBounds;
-      if (!FuncProtoTy->hasParamAnnots())
+      }
+      if (!FuncProtoTy->hasParamAnnots()) {
+        TraverseChildren(E, CSS, Facts, SE);
+        return ResultBounds;
+      }
+      if (SE == SideEffects::Disabled) {
+        TraverseChildren(E, CSS, Facts, SE);
         return ResultBounds;
+      }
+
+      // Perform checking of bounds declarations, if enabled.
+
+      // Recursively traverse the callee.  The arguments will be
+      // traversed below.  This prevents TraverseStmt from
+      // needing to traverse the children of call expressions.
+      TraverseStmt(E->getCallee(), CSS, Facts, SE);
 
       unsigned NumParams = FuncProtoTy->getNumParams();
       unsigned NumArgs = E->getNumArgs();
       unsigned Count = (NumParams < NumArgs) ? NumParams : NumArgs;
       ArrayRef<Expr *> ArgExprs = llvm::makeArrayRef(const_cast<Expr**>(E->getArgs()), E->getNumArgs());
```

Comment:
You need to handle the case where the number of arguments > the number of parameters at the function call.  The additional arguments still need to be traversed.   This can happen at varargs calls, for example.   The additional arguments could require runtime bounds checks within them.  They could also have casts that need to be statically checked (for example, to _Ptr type).